src/NServiceBus.Core/Transports/Learning/LearningTransportDispatcher.cs

```diff
@@ -0,0 +1,170 @@
+namespace NServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using DelayedDelivery;
+    using DeliveryConstraints;
+    using Extensibility;
+    using Transport;
+
+    class LearningTransportDispatcher : IDispatchMessages
+    {
+        public LearningTransportDispatcher(string basePath)
+        {
+            this.basePath = basePath;
+        }
+
+        public Task Dispatch(TransportOperations outgoingMessages, TransportTransaction transaction, ContextBag context)
+        {
+            return Task.WhenAll(
+                DispatchUnicast(outgoingMessages.UnicastTransportOperations, transaction),
+                DispatchMulticast(outgoingMessages.MulticastTransportOperations, transaction));
+        }
+
+        async Task DispatchMulticast(IEnumerable<MulticastTransportOperation> transportOperations, TransportTransaction transaction)
+        {
+            var tasks = new List<Task>();
+            foreach (var transportOperation in transportOperations)
+            {
+                var subscribers = await GetSubscribersFor(transportOperation.MessageType)
+                    .ConfigureAwait(false);
+
+                foreach (var subscriber in subscribers)
+                {
+                    tasks.Add(WriteMessage(subscriber, transportOperation, transaction));
+                }
+            }
+            await Task.WhenAll(tasks)
+                .ConfigureAwait(false);
+        }
+
+
+        Task DispatchUnicast(IEnumerable<UnicastTransportOperation> operations, TransportTransaction transaction)
+        {
+            return Task.WhenAll(operations.Select(operation => WriteMessage(operation.Destination, operation, transaction)));
+        }
+
+        async Task WriteMessage(string destination, IOutgoingTransportOperation transportOperation, TransportTransaction transaction)
+        {
+            var nativeMessageId = Guid.NewGuid().ToString();
+            var destinationPath = Path.Combine(basePath, destination);
+            var bodyDir = Path.Combine(destinationPath, ".bodies");
+
+            Directory.CreateDirectory(bodyDir);
+
+            var bodyPath = Path.Combine(bodyDir, nativeMessageId) + ".txt";
+
+            await AsyncFile.WriteBytes(bodyPath, transportOperation.Message.Body).ConfigureAwait(false);
+
+            var messageContents = $"{bodyPath}{Environment.NewLine}{HeaderSerializer.Serialize(transportOperation.Message.Headers)}";
+
+            DateTime? timeToDeliver = null;
+            DelayDeliveryWith delayDeliveryWith;
+
+            if (transportOperation.DeliveryConstraints.TryGet(out delayDeliveryWith))
+            {
+                timeToDeliver = DateTime.UtcNow + delayDeliveryWith.Delay;
```

Comment:
Potentially a non-issue, just thinking out loud. Since we write file per file the timeToDeliver using DateTime.UtcNow would also move in time during sending of multiple messages when the IO operations take longer. Is that an issue or not?